client/sites/settings/v2/router.tsx

```diff
@@ -0,0 +1,150 @@
+	path: 'site-visibility',
+	loader: ( { params: { siteSlug } } ) =>
+		queryClient.ensureQueryData( siteSettingsQuery( siteSlug ) ),
+} ).lazy( () =>
+	import( 'calypso/dashboard/sites/settings-site-visibility' ).then( ( d ) =>
+		createLazyRoute( 'site-visibility' )( {
+			component: () => <d.default siteSlug={ siteRoute.useParams().siteSlug } />,
+		} )
+	)
+);
+
+const subscriptionGiftingRoute = createRoute( {
+	getParentRoute: () => siteRoute,
+	path: 'subscription-gifting',
+	loader: ( { params: { siteSlug } } ) =>
+		queryClient.ensureQueryData( siteSettingsQuery( siteSlug ) ),
+} ).lazy( () =>
+	import( 'calypso/dashboard/sites/settings-subscription-gifting' ).then( ( d ) =>
+		createLazyRoute( 'subscription-gifting' )( {
+			component: () => <d.default siteSlug={ siteRoute.useParams().siteSlug } />,
+		} )
+	)
+);
+
+const createRouteTree = () =>
+	rootRoute.addChildren( [
+		siteRoute.addChildren( [ settingsRoute, siteVisibilityRoute, subscriptionGiftingRoute ] ),
+		dashboardSiteSettingsCompatibilityRouteRoot,
+		dashboardSiteSettingsCompatibilityRouteWithFeature,
+	] );
+
+const routeTree = createRouteTree();
+
+const isCompatibilityRoute = ( router: Router< typeof routeTree >, url: string ) => {
+	const matches = router.matchRoutes( url );
+	if ( ! matches ) {
+		return false;
+	}
+
+	return matches.some(
+		( match: { routeId: string } ) =>
+			match.routeId === dashboardSiteSettingsCompatibilityRouteRoot.id ||
+			match.routeId === dashboardSiteSettingsCompatibilityRouteWithFeature.id
+	);
+};
+
+const syncMemoryRouterToBrowserHistory = ( router: Router< typeof routeTree > ) => {
+	let lastPath = '';
+
+	// Sync TanStack Router's history to the browser history.
+	router.history.subscribe( () => {
+		const { pathname, search } = router.history.location;
+		const newUrl = `${ pathname }${ search }`;
+
+		// Avoid pushing redirect routes to the browser history.
+		if ( isCompatibilityRoute( router, newUrl ) ) {
+			return;
+		}
+
+		if ( window.location.pathname + window.location.search !== newUrl ) {
+			window.history.pushState( null, '', newUrl );
+			lastPath = newUrl;
+		}
+	} );
+
+	window.addEventListener( 'popstate', () => {
+		const currentPath = `${ window.location.pathname }${ window.location.search }`;
+		const basepath = router.options.basepath;
+
+		// Avoid handling routes outside of the basepath.
+		if ( basepath && ! currentPath.startsWith( basepath ) ) {
+			return;
+		}
+
+		if ( currentPath !== lastPath ) {
+			router.navigate( { to: currentPath, replace: true } );
+			lastPath = currentPath;
+		}
+	} );
+};
+
+export const getRouter = () => {
+	const router = new Router( {
+		routeTree,
+		basepath: '/sites/settings/v2',
+		defaultPreload: 'intent',
+		defaultPreloadStaleTime: 0,
+		defaultNotFoundComponent: () => null,
+
+		// Use memory history to compartmentize TanStack Router's history management.
+		// This is necessary in order to not pollute the browser history which is used by page.js
```

Comment:
This comment threw me off initially. Because we _do_ sort of pollute the browser history with the Tan stack navigations. Its just that we don't want Tan stack to do that "polluting", we want to do it manually ourselves.

Reply:
Updated the comment to be more about keeping TanStack's history implementation separated from the browser history.